File: scripts/programmers_catalog_worker.py
#!/usr/bin/env python3
"""
Fetches Programmers catalog data and writes a normalized JSON snapshot into
PROGRAMMERS_IMPORT_DIR for the backend file source to ingest.

Priority:
1. Crawl Programmers challenge pages and extract lesson data.
2. Fallback to a JSON feed URL if crawling yields no items.
3. Optionally trigger the backend admin ingest endpoint after writing.
"""

from __future__ import annotations

import argparse
import html
import json
import os
import re
import sys
import time
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterable
from urllib import error, parse, request


USER_AGENT = "ctps-programmers-catalog-worker/1.0"
DEFAULT_URL_TEMPLATE = "https://school.programmers.co.kr/learn/challenges?page={page}"


@dataclass
class CatalogItem:
    externalId: str
    title: str
    problemNumber: str
    difficulty: str
    tags: list[str]
    externalUrl: str
    recommendationReason: str


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Collect Programmers catalog items and write a normalized import JSON."
    )
    parser.add_argument("--import-dir", default=os.getenv("PROGRAMMERS_IMPORT_DIR", ""))
    parser.add_argument("--pages", type=int, default=int(os.getenv("PROGRAMMERS_CRAWLER_PAGES", "3")))
    parser.add_argument(
        "--url-template",
        default=os.getenv("PROGRAMMERS_CRAWLER_URL_TEMPLATE", DEFAULT_URL_TEMPLATE),
        help="Listing URL template. Use {page} placeholder.",
    )
    parser.add_argument(
        "--fallback-feed-url",
        default=os.getenv("PROGRAMMERS_FEED_URL", ""),
        help="Fallback JSON feed URL used only when crawl result is empty.",
    )
    parser.add_argument("--delay-ms", type=int, default=int(os.getenv("PROGRAMMERS_CRAWLER_DELAY_MS", "250")))
    parser.add_argument("--retention", type=int, default=int(os.getenv("PROGRAMMERS_CRAWLER_RETENTION", "10")))
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument(
        "--ingest-url",
        default=os.getenv("CTPS_PROGRAMMERS_INGEST_URL", ""),
        help="Optional admin ingest endpoint, for example http://localhost:8080/api/admin/external-search/programmers/ingest",
    )
    parser.add_argument("--admin-token", default=os.getenv("ADMIN_SECURITY_TOKEN", ""))
    return parser.parse_args()


def fetch_text(url: str) -> str:
    req = request.Request(url, headers={"User-Agent": USER_AGENT})
    with request.urlopen(req, timeout=20) as response:
        encoding = response.headers.get_content_charset() or "utf-8"
        return response.read().decode(encoding, errors="replace")


def fetch_json(url: str) -> Any:
    return json.loads(fetch_text(url))


def normalize_difficulty(raw: Any) -> str:
    if raw is None:
        return "medium"

    text = str(raw).strip().lower()
    if text in {"1", "lv1", "level1", "easy"}:
        return "easy"
    if text in {"2", "3", "lv2", "lv3", "level2", "level3", "medium"}:
        return "medium"
    if text in {"4", "5", "lv4", "lv5", "level4", "level5", "hard"}:
        return "hard"

    if "easy" in text:
        return "easy"
    if "hard" in text:
        return "hard"
    if "medium" in text:
        return "medium"

    digits = re.findall(r"\d+", text)
    if digits:
        level = int(digits[0])
        if level <= 1:
            return "easy"
        if level <= 3:
            return "medium"
        return "hard"

    return "medium"


def strip_tags(value: str) -> str:
    without_tags = re.sub(r"<[^>]+>", " ", value)
    return html.unescape(re.sub(r"\s+", " ", without_tags)).strip()


def first_present(data: dict[str, Any], keys: Iterable[str]) -> Any:
    for key in keys:
        value = data.get(key)
        if value not in (None, "", [], {}):
            return value
    return None


def normalize_tags(raw: Any) -> list[str]:
    if raw is None:
        return []
    if isinstance(raw, str):
        return [raw.strip()] if raw.strip() else []
    if isinstance(raw, list):
        tags: list[str] = []
        for item in raw:
            if isinstance(item, str) and item.strip():
                tags.append(item.strip())
            elif isinstance(item, dict):
                tag = first_present(item, ("name", "label", "title", "value"))
                if isinstance(tag, str) and tag.strip():
                    tags.append(tag.strip())
        return list(dict.fromkeys(tags))
    return []


def normalize_candidate(candidate: dict[str, Any]) -> CatalogItem | None:
    identifier = first_present(candidate, ("externalId", "lessonId", "lesson_id", "problemId", "id"))
    title = first_present(candidate, ("title", "name", "lessonTitle"))
    if not identifier or not title:
        return None

    problem_number = str(first_present(candidate, ("problemNumber", "problem_number", "id", "lessonId", "lesson_id")) or identifier)
    external_url = first_present(candidate, ("externalUrl", "url", "link"))
    if isinstance(external_url, str) and external_url.startswith("/"):
        external_url = parse.urljoin("https://school.programmers.co.kr", external_url)
    if not external_url:
        external_url = f"https://school.programmers.co.kr/learn/courses/30/lessons/{identifier}"

    tags = normalize_tags(first_present(candidate, ("tags", "skills", "categories", "categoryNames")))
    recommendation_reason = first_present(
        candidate,
        ("recommendationReason", "recommendation_reason", "summary", "description"),
    )
    if not recommendation_reason:
        recommendation_reason = "프로그래머스에서 탐색 가능한 추천 문제입니다."

    return CatalogItem(
        externalId=f"programmers-{identifier}",
        title=str(title).strip(),
        problemNumber=problem_number,
        difficulty=normalize_difficulty(first_present(candidate, ("difficulty", "level", "difficultyLevel"))),
        tags=tags,
        externalUrl=str(external_url),
        recommendationReason=str(recommendation_reason).strip(),
    )


def walk_candidates(node: Any) -> Iterable[dict[str, Any]]:
    if isinstance(node, dict):
        yield node
        for value in node.values():
            yield from walk_candidates(value)
    elif isinstance(node, list):
        for item in node:
            yield from walk_candidates(item)


def extract_next_data(html_text: str) -> Any | None:
    match = re.search(
        r'<script[^>]+id="__NEXT_DATA__"[^>]*>\s*(.*?)\s*</script>',
        html_text,
        re.DOTALL | re.IGNORECASE,
    )
    if not match:
        return None
    try:
        return json.loads(match.group(1))
    except json.JSONDecodeError:
        return None


def extract_anchor_fallback(html_text: str) -> list[CatalogItem]:
    items: list[CatalogItem] = []
    pattern = re.compile(
        r'<a[^>]+href="(?P<href>/learn/courses/\d+/lessons/(?P<id>\d+))"[^>]*>(?P<content>.*?)</a>',
        re.DOTALL | re.IGNORECASE,
    )
    for match in pattern.finditer(html_text):
        title = strip_tags(match.group("content"))
        if not title:
            continue
        items.append(
            CatalogItem(
                externalId=f"programmers-{match.group('id')}",
                title=title,
                problemNumber=match.group("id"),
                difficulty="medium",
                tags=[],
                externalUrl=parse.urljoin("https://school.programmers.co.kr", match.group("href")),
                recommendationReason="프로그래머스 공개 문제 목록에서 수집한 문제입니다.",
            )
        )
    return items


def crawl_programmers_catalog(url_template: str, pages: int, delay_ms: int) -> list[CatalogItem]:
    deduped: dict[str, CatalogItem] = {}

    for page in range(1, pages + 1):
        url = url_template.format(page=page)
        html_text = fetch_text(url)

        next_data = extract_next_data(html_text)
        if next_data is not None:
            for candidate in walk_candidates(next_data):
                item = normalize_candidate(candidate)
                if item:
                    deduped[item.externalId] = item

        if not deduped:
            for item in extract_anchor_fallback(html_text):
                deduped[item.externalId] = item

        if page < pages and delay_ms > 0:
            time.sleep(delay_ms / 1000)

    return list(deduped.values())


def load_fallback_feed(feed_url: str) -> list[CatalogItem]:
    if not feed_url:
        return []
    payload = fetch_json(feed_url)
    if not isinstance(payload, list):
        return []

    items: list[CatalogItem] = []
    for entry in payload:
        if not isinstance(entry, dict):
            continue
        item = normalize_candidate(entry)
        if item:
            items.append(item)
    return items


def write_snapshot(import_dir: str, items: list[CatalogItem], retention: int) -> Path:
    target_dir = Path(import_dir)
    target_dir.mkdir(parents=True, exist_ok=True)

    timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    target = target_dir / f"programmers-catalog-{timestamp}.json"
    temp = target_dir / f".programmers-catalog-{timestamp}.tmp"

    payload = [asdict(item) for item in items]
    temp.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")
    temp.replace(target)

    snapshots = sorted(
        (path for path in target_dir.glob("programmers-catalog-*.json") if path.is_file()),
        key=lambda path: path.stat().st_mtime,
        reverse=True,
    )
    for old_file in snapshots[retention:]:
        old_file.unlink(missing_ok=True)

    return target


def trigger_ingest(ingest_url: str, admin_token: str) -> None:
    if not ingest_url:
        return

    headers = {"User-Agent": USER_AGENT}
    if admin_token:
        headers["X-Admin-Token"] = admin_token

    req = request.Request(ingest_url, method="POST", headers=headers)
    with request.urlopen(req, timeout=20) as response:
        if response.status >= 400:
            raise RuntimeError(f"ingest trigger failed with status={response.status}")


def main() -> int:
    args = parse_args()
    if not args.import_dir and not args.dry_run:
        print("PROGRAMMERS_IMPORT_DIR 또는 --import-dir 설정이 필요합니다.", file=sys.stderr)
        return 1

    try:
        items = crawl_programmers_catalog(args.url_template, args.pages, args.delay_ms)
        source_name = "crawl"
        if not items:
            items = load_fallback_feed(args.fallback_feed_url)
            source_name = "fallback-feed"

        if not items:
            print("수집된 프로그래머스 카탈로그 항목이 없습니다.", file=sys.stderr)
            return 2

        if args.dry_run:
            print(json.dumps([asdict(item) for item in items[:5]], ensure_ascii=False, indent=2))
            print(f"dry-run source={source_name} count={len(items)}")
            return 0

        output_file = write_snapshot(args.import_dir, items, args.retention)
        trigger_ingest(args.ingest_url, args.admin_token)
        print(f"wrote {len(items)} items from {source_name} to {output_file}")
        return 0
    except error.HTTPError as http_error:
        print(f"http error: {http_error.code} {http_error.reason}", file=sys.stderr)
        return 3
    except Exception as exception:
        print(f"worker failed: {exception}", file=sys.stderr)
        return 4


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/programmers_catalog_worker.mjs
import fs from 'node:fs/promises';
import path from 'node:path';
import process from 'node:process';
import { chromium } from 'playwright';

const DEFAULT_URL_TEMPLATE = 'https://school.programmers.co.kr/learn/challenges?page={page}';
const DEFAULT_RETENTION = 10;
const DEFAULT_DELAY_MS = 250;
const DEFAULT_PAGES = 3;
const DEFAULT_MODE = 'incremental';

function parseArgs(argv) {
  const options = {
    importDir: process.env.PROGRAMMERS_IMPORT_DIR ?? '',
    pages: Number.parseInt(process.env.PROGRAMMERS_CRAWLER_PAGES ?? String(DEFAULT_PAGES), 10),
    urlTemplate: process.env.PROGRAMMERS_CRAWLER_URL_TEMPLATE ?? DEFAULT_URL_TEMPLATE,
    delayMs: Number.parseInt(process.env.PROGRAMMERS_CRAWLER_DELAY_MS ?? String(DEFAULT_DELAY_MS), 10),
    retention: Number.parseInt(process.env.PROGRAMMERS_CRAWLER_RETENTION ?? String(DEFAULT_RETENTION), 10),
    ingestUrl: process.env.CTPS_PROGRAMMERS_INGEST_URL ?? '',
    adminToken: process.env.ADMIN_SECURITY_TOKEN ?? '',
    mode: process.env.PROGRAMMERS_CRAWLER_MODE ?? DEFAULT_MODE,
    dryRun: false,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];
    const next = argv[index + 1];
    switch (arg) {
      case '--import-dir':
        options.importDir = next ?? '';
        index += 1;
        break;
      case '--pages':
        options.pages = Number.parseInt(next ?? String(DEFAULT_PAGES), 10);
        index += 1;
        break;
      case '--url-template':
        options.urlTemplate = next ?? DEFAULT_URL_TEMPLATE;
        index += 1;
        break;
      case '--delay-ms':
        options.delayMs = Number.parseInt(next ?? String(DEFAULT_DELAY_MS), 10);
        index += 1;
        break;
      case '--retention':
        options.retention = Number.parseInt(next ?? String(DEFAULT_RETENTION), 10);
        index += 1;
        break;
      case '--ingest-url':
        options.ingestUrl = next ?? '';
        index += 1;
        break;
      case '--admin-token':
        options.adminToken = next ?? '';
        index += 1;
        break;
      case '--mode':
        options.mode = next ?? DEFAULT_MODE;
        index += 1;
        break;
      case '--full-refresh':
        options.mode = 'full-refresh';
        break;
      case '--dry-run':
        options.dryRun = true;
        break;
      default:
        break;
    }
  }

  if (!Number.isFinite(options.pages) || options.pages <= 0) {
    options.pages = DEFAULT_PAGES;
  }
  if (!Number.isFinite(options.delayMs) || options.delayMs < 0) {
    options.delayMs = DEFAULT_DELAY_MS;
  }
  if (!Number.isFinite(options.retention) || options.retention <= 0) {
    options.retention = DEFAULT_RETENTION;
  }
  if (!['incremental', 'full-refresh'].includes(options.mode)) {
    options.mode = DEFAULT_MODE;
  }

  return options;
}

function wait(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function normalizeWhitespace(value) {
  return String(value ?? '')
    .replace(/\s+/g, ' ')
    .trim();
}

function normalizeDifficulty(rawValue) {
  const text = normalizeWhitespace(rawValue).toLowerCase();
  if (!text) return 'medium';
  const levelMatch = text.match(/(?:lv|level|레벨|난이도)\s*\.?\s*(\d)/i);
  if (levelMatch) {
    const level = Number.parseInt(levelMatch[1], 10);
    if (level <= 1) return 'easy';
    if (level <= 3) return 'medium';
    return 'hard';
  }
  if (text.includes('lv. 0') || text.includes('level 0') || text.includes('easy')) return 'easy';
  if (text.includes('lv. 1') || text.includes('level 1')) return 'easy';
  if (
    text.includes('lv. 4')
    || text.includes('lv. 5')
    || text.includes('level 4')
    || text.includes('level 5')
    || text.includes('hard')
  ) {
    return 'hard';
  }
  return 'medium';
}

function expandTags(partText) {
  const cleaned = normalizeWhitespace(partText);
  if (!cleaned) return [];

  const tags = new Set(
    cleaned
      .split(/[|,>]/)
      .flatMap((part) => part.split('/'))
      .map((part) => normalizeWhitespace(part))
      .filter(Boolean)
  );
  tags.add(cleaned);
  const aliasMatches = cleaned.match(/\(([^)]+)\)/g) ?? [];
  for (const aliasGroup of aliasMatches) {
    const aliases = aliasGroup
      .replace(/[()]/g, '')
      .split('/')
      .map((value) => normalizeWhitespace(value))
      .filter(Boolean);
    for (const alias of aliases) {
      tags.add(alias);
    }
  }
  return [...tags];
}

function firstPresent(candidate, keys) {
  for (const key of keys) {
    const value = candidate?.[key];
    if (value !== null && value !== undefined && value !== '' && (!Array.isArray(value) || value.length > 0)) {
      return value;
    }
  }
  return null;
}

function* walkCandidates(node) {
  if (Array.isArray(node)) {
    for (const item of node) {
      yield* walkCandidates(item);
    }
    return;
  }
  if (node && typeof node === 'object') {
    yield node;
    for (const value of Object.values(node)) {
      yield* walkCandidates(value);
    }
  }
}

function normalizeEmbeddedTags(rawValue) {
  if (!rawValue) return [];
  if (typeof rawValue === 'string') {
    return expandTags(rawValue);
  }
  if (Array.isArray(rawValue)) {
    return [...new Set(rawValue.flatMap((entry) => {
      if (typeof entry === 'string') return expandTags(entry);
      if (entry && typeof entry === 'object') {
        const value = firstPresent(entry, ['name', 'label', 'title', 'value']);
        return typeof value === 'string' ? expandTags(value) : [];
      }
      return [];
    }))];
  }
  return [];
}

async function extractEmbeddedLessonMetadata(page, lesson) {
  const nextDataText = await page.locator('script#__NEXT_DATA__').textContent().catch(() => '');
  if (!nextDataText) {
    return { title: '', difficulty: '', tags: [] };
  }

  try {
    const payload = JSON.parse(nextDataText);
    const resolvedTags = new Set();
    let resolvedTitle = '';
    let resolvedDifficulty = '';

    for (const candidate of walkCandidates(payload)) {
      const identifier = String(firstPresent(candidate, ['lessonId', 'lesson_id', 'problemId', 'id']) ?? '');
      const candidateTitle = normalizeWhitespace(firstPresent(candidate, ['title', 'name', 'lessonTitle']) ?? '');
      const sameLesson = identifier === lesson.lessonId || (!!candidateTitle && candidateTitle === lesson.title);
      if (!sameLesson) {
        continue;
      }

      if (!resolvedTitle && candidateTitle) {
        resolvedTitle = candidateTitle;
      }
      if (!resolvedDifficulty) {
        const difficulty = firstPresent(candidate, ['difficulty', 'level', 'difficultyLevel']);
        if (difficulty !== null && difficulty !== undefined) {
          resolvedDifficulty = String(difficulty);
        }
      }

      normalizeEmbeddedTags(firstPresent(candidate, ['tags', 'skills', 'categories', 'categoryNames']))
        .forEach((tag) => resolvedTags.add(tag));
    }

    return {
      title: resolvedTitle,
      difficulty: resolvedDifficulty,
      tags: [...resolvedTags],
    };
  } catch {
    return { title: '', difficulty: '', tags: [] };
  }
}

function buildExternalId(lessonId) {
  return `programmers-${lessonId}`;
}

function resolveAbsoluteUrl(href) {
  return new URL(href, 'https://school.programmers.co.kr').toString();
}

async function collectLessonLinks(page, urlTemplate, pages, delayMs) {
  const lessons = new Map();

  for (let currentPage = 1; currentPage <= pages; currentPage += 1) {
    const url = urlTemplate.replace('{page}', String(currentPage));
    await page.goto(url, { waitUntil: 'networkidle', timeout: 60000 });

    const entries = await page.locator('a[href*="/learn/courses/"][href*="/lessons/"]').evaluateAll((anchors) =>
      anchors
        .map((anchor) => ({
          href: anchor.getAttribute('href') ?? '',
          title: (anchor.textContent ?? '').trim(),
        }))
        .filter((entry) => entry.href && entry.title)
    );

    for (const entry of entries) {
      const lessonIdMatch = entry.href.match(/lessons\/(\d+)/);
      if (!lessonIdMatch) {
        continue;
      }

      const lessonId = lessonIdMatch[1];
      if (!lessons.has(lessonId)) {
        lessons.set(lessonId, {
          lessonId,
          title: normalizeWhitespace(entry.title),
          externalUrl: resolveAbsoluteUrl(entry.href),
        });
      }
    }

    if (currentPage < pages && delayMs > 0) {
      await wait(delayMs);
    }
  }

  return [...lessons.values()];
}

async function readLatestSnapshot(importDir) {
  if (!importDir) {
    return [];
  }

  try {
    const files = (await fs.readdir(importDir))
      .filter((file) => file.startsWith('programmers-catalog-') && file.endsWith('.json'))
      .map((file) => path.join(importDir, file));

    if (files.length === 0) {
      return [];
    }

    const statEntries = await Promise.all(
      files.map(async (file) => ({
        file,
        mtimeMs: (await fs.stat(file)).mtimeMs,
      }))
    );

    const latestFile = statEntries.sort((left, right) => right.mtimeMs - left.mtimeMs)[0]?.file;
    if (!latestFile) {
      return [];
    }

    const raw = await fs.readFile(latestFile, 'utf8');
    const parsed = JSON.parse(raw);
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

async function enrichLesson(page, lesson) {
  await page.goto(lesson.externalUrl, { waitUntil: 'networkidle', timeout: 60000 });
  const embeddedMetadata = await extractEmbeddedLessonMetadata(page, lesson);

  const title =
    normalizeWhitespace(await page.locator('.challenge-title').first().textContent().catch(() => embeddedMetadata.title || lesson.title))
    || embeddedMetadata.title
    || lesson.title;

  const partTexts = await page.locator('a[href*="/learn/courses/30/parts/"]').evaluateAll((anchors) =>
    anchors.map((anchor) => (anchor.textContent ?? '').trim()).filter(Boolean)
  );

  const difficultyTexts = await page
    .locator('text=/Lv\\.|레벨|난이도/i')
    .evaluateAll((nodes) => nodes.map((node) => (node.textContent ?? '').trim()).filter(Boolean))
    .catch(() => []);

  const tags = [...new Set([...embeddedMetadata.tags, ...partTexts.flatMap(expandTags)])];
  const difficulty = normalizeDifficulty(`${embeddedMetadata.difficulty} ${difficultyTexts.join(' ')}`);

  return {
    externalId: buildExternalId(lesson.lessonId),
    title,
    problemNumber: lesson.lessonId,
    difficulty,
    tags,
    externalUrl: lesson.externalUrl,
    recommendationReason: tags.length > 0
      ? `${tags[0]} 유형의 프로그래머스 문제입니다.`
      : '프로그래머스 코딩테스트 연습 문제입니다.',
  };
}

async function collectCatalogItems(options) {
  const existingItems = options.mode === 'incremental'
    ? await readLatestSnapshot(options.importDir)
    : [];
  const existingItemMap = new Map(
    existingItems
      .filter((item) => item && typeof item === 'object' && typeof item.externalId === 'string')
      .map((item) => [item.externalId, item])
  );

  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    locale: 'ko-KR',
    timezoneId: 'Asia/Seoul',
    userAgent:
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 '
      + '(KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36',
  });
  const page = await context.newPage();
  const detailPage = await context.newPage();

  try {
    const lessons = await collectLessonLinks(page, options.urlTemplate, options.pages, options.delayMs);
    const newItems = [];

    for (const lesson of lessons) {
      const externalId = buildExternalId(lesson.lessonId);
      if (options.mode === 'incremental' && existingItemMap.has(externalId)) {
        continue;
      }

      const item = await enrichLesson(detailPage, lesson);
      newItems.push(item);
      if (options.delayMs > 0) {
        await wait(options.delayMs);
      }
    }

    const mergedItems = options.mode === 'incremental'
      ? [
          ...existingItems,
          ...newItems,
        ]
      : newItems;

    return {
      items: mergedItems,
      newItemCount: newItems.length,
      existingItemCount: existingItemMap.size,
    };
  } finally {
    await context.close();
    await browser.close();
  }
}

async function writeSnapshot(importDir, items, retention) {
  await fs.mkdir(importDir, { recursive: true });
  const timestamp = new Date().toISOString().replace(/[-:]/g, '').replace(/\.\d{3}Z$/, 'Z');
  const target = path.join(importDir, `programmers-catalog-${timestamp}.json`);
  const temp = path.join(importDir, `.programmers-catalog-${timestamp}.tmp`);

  await fs.writeFile(temp, `${JSON.stringify(items, null, 2)}\n`, 'utf8');
  await fs.rename(temp, target);

  const files = (await fs.readdir(importDir))
    .filter((file) => file.startsWith('programmers-catalog-') && file.endsWith('.json'))
    .map((file) => path.join(importDir, file));

  const statEntries = await Promise.all(
    files.map(async (file) => ({
      file,
      mtimeMs: (await fs.stat(file)).mtimeMs,
    }))
  );

  for (const { file } of statEntries
    .sort((left, right) => right.mtimeMs - left.mtimeMs)
    .slice(retention)) {
    await fs.rm(file, { force: true });
  }

  return target;
}

async function triggerIngest(ingestUrl, adminToken) {
  if (!ingestUrl) {
    return;
  }

  const response = await fetch(ingestUrl, {
    method: 'POST',
    headers: {
      'User-Agent': 'ctps-programmers-catalog-worker/1.0',
      ...(adminToken ? { 'X-Admin-Token': adminToken } : {}),
    },
  });

  if (!response.ok) {
    throw new Error(`ingest trigger failed with status=${response.status}`);
  }
}

async function main() {
  const options = parseArgs(process.argv.slice(2));
  if (!options.importDir && !options.dryRun) {
    console.error('PROGRAMMERS_IMPORT_DIR 또는 --import-dir 설정이 필요합니다.');
    process.exitCode = 1;
    return;
  }

  const { items, newItemCount, existingItemCount } = await collectCatalogItems(options);
  if (items.length === 0) {
    console.error('수집된 프로그래머스 카탈로그 항목이 없습니다.');
    process.exitCode = 2;
    return;
  }

  if (options.dryRun) {
    console.log(JSON.stringify(items.slice(0, 5), null, 2));
    console.log(
      `dry-run source=playwright mode=${options.mode} total=${items.length} new=${newItemCount} existing=${existingItemCount}`
    );
    return;
  }

  const outputFile = await writeSnapshot(options.importDir, items, options.retention);
  await triggerIngest(options.ingestUrl, options.adminToken);
  console.log(
    `wrote ${items.length} items from playwright to ${outputFile} `
    + `(mode=${options.mode}, new=${newItemCount}, existing=${existingItemCount})`
  );
}

main().catch((error) => {
  console.error(`worker failed: ${error instanceof Error ? error.message : String(error)}`);
  process.exitCode = 4;
});
